src/cargo/core/gc.rs

```diff
@@ -0,0 +1,550 @@
+    config_span: &str,
+) -> CargoResult<Option<Duration>> {
+    let config_span = parse_time_span_for_config(config_name, config_span)?;
+    Ok(Some(maybe_newer_span(config_span, cur_span)))
+}
+
+/// Returns whichever [`Duration`] is shorter.
+fn maybe_newer_span(a: Duration, b: Option<Duration>) -> Duration {
+    match b {
+        Some(b) => {
+            if b < a {
+                b
+            } else {
+                a
+            }
+        }
+        None => a,
+    }
+}
+
+/// Parses a frequency string.
+///
+/// Returns `Ok(None)` if the frequency is "never".
+fn parse_frequency(frequency: &str) -> CargoResult<Option<Duration>> {
+    if frequency == "always" {
+        return Ok(Some(Duration::new(0, 0)));
+    } else if frequency == "never" {
+        return Ok(None);
+    }
+    let duration = maybe_parse_time_span(frequency).ok_or_else(|| {
+        format_err!(
+            "config option `gc.auto.frequency` expected a value of \"always\", \"never\", \
+             or \"N seconds/minutes/days/weeks/months\", got: {frequency:?}"
+        )
+    })?;
+    Ok(Some(duration))
+}
+
+/// Parses a time span value fetched from config.
+///
+/// This is here to provide better error messages specific to reading from
+/// config.
+fn parse_time_span_for_config(config_name: &str, span: &str) -> CargoResult<Duration> {
+    maybe_parse_time_span(span).ok_or_else(|| {
+        format_err!(
+            "config option `{config_name}` expected a value of the form \
+             \"N seconds/minutes/days/weeks/months\", got: {span:?}"
+        )
+    })
+}
+
+/// Parses a time span string.
+///
+/// Returns None if the value is not valid. See [`parse_time_span`] if you
+/// need a variant that generates an error message.
+fn maybe_parse_time_span(span: &str) -> Option<Duration> {
+    let Some(right_i) = span.find(|c: char| !c.is_ascii_digit()) else {
+        return None;
+    };
+    let left = &span[..right_i];
+    let mut right = &span[right_i..];
+    if right.starts_with(' ') {
+        right = &right[1..];
+    }
+    let count: u64 = left.parse().ok()?;
+    let factor = match right {
+        "second" | "seconds" => 1,
+        "minute" | "minutes" => 60,
+        "hour" | "hours" => 60 * 60,
+        "day" | "days" => 24 * 60 * 60,
+        "week" | "weeks" => 7 * 24 * 60 * 60,
+        "month" | "months" => 30 * 24 * 60 * 60,
+        _ => return None,
+    };
+    Some(Duration::from_secs(factor * count))
+}
+
+/// Parses a time span string.
+pub fn parse_time_span(span: &str) -> CargoResult<Duration> {
+    maybe_parse_time_span(span).ok_or_else(|| {
+        format_err!(
+            "expected a value of the form \
+             \"N seconds/minutes/days/weeks/months\", got: {span:?}"
+        )
+    })
+}
+
+/// Parses a file size using metric or IEC units.
+pub fn parse_human_size(input: &str) -> CargoResult<u64> {
+    let re = regex::Regex::new(r"(?i)^([0-9]+(\.[0-9])?) ?(b|kb|mb|gb|kib|mib|gib)?$").unwrap();
```

Comment:
How come the mix of regex (for bytes) and hand parsing (time)?

Reply:
The split on time was simpler (since it requires a unit, and doesn't allow fractions), so it seemed more straightforward to me to just split the string and parse the left/right sides. I could use a regex there, too, if you want, but I think overall it would be more lines of code. The reason the byte size uses a regex is because it was less code and more straightforward compared to trying to do it all with just str methods.